File: api/telegram-webhook.py
"""
Vercel Serverless Function for Telegram Webhook API
This handles Telegram notifications for the UmraGO Team Planner

Deploy this to Vercel as serverless functions
"""

import os
import json
from http.server import BaseHTTPRequestHandler
from supabase import create_client, Client
import requests
import logging

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

# Configuration from environment variables
TELEGRAM_BOT_TOKEN = os.getenv('TELEGRAM_BOT_TOKEN', '')
SUPABASE_URL = os.getenv('SUPABASE_URL', '')
SUPABASE_KEY = os.getenv('SUPABASE_ANON_KEY', '')
WEB_APP_URL = os.getenv('WEB_APP_URL', 'https://team-planner-task-tracker.vercel.app/')
WEBHOOK_SECRET = os.getenv('WEBHOOK_SECRET', '')

# Initialize Supabase client
supabase: Client = create_client(SUPABASE_URL, SUPABASE_KEY) if SUPABASE_URL and SUPABASE_KEY else None


def send_telegram_message(chat_id: str, message: str):
    """Send a message via Telegram Bot API"""
    try:
        url = f"https://api.telegram.org/bot{TELEGRAM_BOT_TOKEN}/sendMessage"
        payload = {
            'chat_id': chat_id,
            'text': message,
            'parse_mode': 'HTML',
            'disable_web_page_preview': True
        }
        
        response = requests.post(url, json=payload, timeout=10)
        return response.json()
        
    except Exception as e:
        logger.error(f"Error sending Telegram message: {e}")
        return None


def get_active_groups():
    """Get all active Telegram groups"""
    try:
        if not supabase:
            return []
            
        response = supabase.table('telegram_groups')\
            .select('*')\
            .eq('is_active', True)\
            .execute()
        
        return response.data
        
    except Exception as e:
        logger.error(f"Error getting active groups: {e}")
        return []


def log_notification(notification_type: str, task_id: str, chat_id: str, message: str, status: str = 'sent', error: str = None):
    """Log notification to database"""
    try:
        if not supabase:
            return
            
        data = {
            'notification_type': notification_type,
            'task_id': task_id,
            'chat_id': chat_id,
            'message_text': message,
            'status': status
        }
        
        if error:
            data['error_message'] = error
        
        supabase.table('telegram_notifications').insert(data).execute()
        
    except Exception as e:
        logger.error(f"Error logging notification: {e}")


class handler(BaseHTTPRequestHandler):
    def do_OPTIONS(self):
        """Handle CORS preflight"""
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'POST, GET, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type, X-Webhook-Secret')
        self.end_headers()
    
    def do_GET(self):
        """Health check endpoint"""
        if self.path == '/api/telegram-webhook/health':
            self.send_response(200)
            self.send_header('Content-Type', 'application/json')
            self.send_header('Access-Control-Allow-Origin', '*')
            self.end_headers()
            self.wfile.write(json.dumps({
                'status': 'ok',
                'service': 'UmraGO Telegram Webhook API'
            }).encode())
        else:
            self.send_response(404)
            self.end_headers()
    
    def do_POST(self):
        """Handle webhook requests"""
        # Set CORS headers
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Content-Type', 'application/json')
        
        try:
            # Read request body
            content_length = int(self.headers.get('Content-Length', 0))
            body = self.rfile.read(content_length)
            data = json.loads(body.decode('utf-8'))
            
            # Verify webhook secret
            secret = self.headers.get('X-Webhook-Secret', '')
            if secret != WEBHOOK_SECRET:
                self.send_response(401)
                self.end_headers()
                self.wfile.write(json.dumps({'error': 'Unauthorized'}).encode())
                return
            
            # Route to appropriate handler based on path
            if '/task-created' in self.path:
                result = self.handle_task_created(data)
            elif '/task-updated' in self.path:
                result = self.handle_task_updated(data)
            elif '/task-completed' in self.path:
                result = self.handle_task_completed(data)
            elif '/comment-added' in self.path:
                result = self.handle_comment_added(data)
            elif '/broadcast' in self.path:
                result = self.handle_broadcast(data)
            elif '/deadline-reminder' in self.path:
                result = self.handle_deadline_reminder(data)
            else:
                self.send_response(404)
                self.end_headers()
                self.wfile.write(json.dumps({'error': 'Endpoint not found'}).encode())
                return
            
            # Send response
            self.send_response(200)
            self.end_headers()
            self.wfile.write(json.dumps(result).encode())
            
        except Exception as e:
            logger.error(f"Error handling request: {e}")
            self.send_response(500)
            self.end_headers()
            self.wfile.write(json.dumps({'error': str(e)}).encode())
    
    def handle_task_created(self, data):
        """Handle task creation webhook"""
        task = data.get('record')
        
        if not task:
            return {'error': 'No task data provided'}
        
        message = (
            "🆕 <b>NEW TASK CREATED</b>\n\n"
            f"📋 <b>Title:</b> {task['title']}\n"
            f"👤 <b>Assigned to:</b> {task['assigned_to']}\n"
            f"📅 <b>Deadline:</b> {task['deadline'][:10]}\n"
            f"⚡ <b>Priority:</b> {task['priority'].upper()}\n"
            f"✍️ <b>Created by:</b> {task['created_by']}\n\n"
            f"👉 <a href='{WEB_APP_URL}'>View Task Details</a>"
        )
        
        groups = get_active_groups()
        sent_count = 0
        failed_count = 0
        
        for group in groups:
            result = send_telegram_message(group['chat_id'], message)
            
            if result and result.get('ok'):
                sent_count += 1
                log_notification('task_created', task['id'], group['chat_id'], message, 'sent')
            else:
                failed_count += 1
                error_msg = result.get('description') if result else 'Unknown error'
                log_notification('task_created', task['id'], group['chat_id'], message, 'failed', error_msg)
        
        return {
            'success': True,
            'sent_to': sent_count,
            'failed': failed_count,
            'total_groups': len(groups)
        }
    
    def handle_task_updated(self, data):
        """Handle task update webhook"""
        task = data.get('record')
        old_task = data.get('old_record')
        
        if not task:
            return {'error': 'No task data provided'}
        
        changes = []
        if old_task:
            if task.get('status') != old_task.get('status'):
                changes.append(f"Status: {old_task.get('status')} → {task.get('status')}")
            if task.get('priority') != old_task.get('priority'):
                changes.append(f"Priority: {old_task.get('priority')} → {task.get('priority')}")
        
        if not changes:
            return {'success': True, 'message': 'No significant changes'}
        
        message = (
            "🔄 <b>TASK UPDATED</b>\n\n"
            f"📋 <b>Task:</b> {task['title']}\n\n"
            f"<b>Changes:</b>\n" + "\n".join([f"• {c}" for c in changes]) + "\n\n"
            f"👉 <a href='{WEB_APP_URL}'>View Task Details</a>"
        )
        
        groups = get_active_groups()
        sent_count = 0
        
        for group in groups:
            result = send_telegram_message(group['chat_id'], message)
            if result and result.get('ok'):
                sent_count += 1
                log_notification('task_updated', task['id'], group['chat_id'], message, 'sent')
        
        return {'success': True, 'sent_to': sent_count}
    
    def handle_task_completed(self, data):
        """Handle task completion webhook"""
        task = data.get('record')
        
        if not task or task.get('status') != 'completed':
            return {'error': 'Invalid task data'}
        
        message = (
            "✅ <b>TASK COMPLETED</b>\n\n"
            f"📋 <b>Task:</b> {task['title']}\n"
            f"👤 <b>Completed by:</b> {task['assigned_to']}\n"
            f"🎉 Great job!\n\n"
            f"👉 <a href='{WEB_APP_URL}'>View Details</a>"
        )
        
        groups = get_active_groups()
        sent_count = 0
        
        for group in groups:
            result = send_telegram_message(group['chat_id'], message)
            if result and result.get('ok'):
                sent_count += 1
                log_notification('task_completed', task['id'], group['chat_id'], message, 'sent')
        
        return {'success': True, 'sent_to': sent_count}
    
    def handle_comment_added(self, data):
        """Handle new comment webhook"""
        comment = data.get('record')
        
        if not comment:
            return {'error': 'No comment data provided'}
        
        try:
            task_response = supabase.table('tasks')\
                .select('*')\
                .eq('id', comment['task_id'])\
                .single()\
                .execute()
            
            if not task_response.data:
                return {'error': 'Task not found'}
            
            task = task_response.data
        except Exception as e:
            logger.error(f"Error fetching task: {e}")
            return {'error': 'Failed to fetch task details'}
        
        message = (
            "💬 <b>NEW COMMENT</b>\n\n"
            f"📋 <b>Task:</b> {task['title']}\n"
            f"👤 <b>From:</b> {comment['user_name']}\n"
            f"💭 <b>Comment:</b> {comment['comment_text']}\n\n"
            f"👉 <a href='{WEB_APP_URL}'>View Task</a>"
        )
        
        groups = get_active_groups()
        sent_count = 0
        
        for group in groups:
            result = send_telegram_message(group['chat_id'], message)
            if result and result.get('ok'):
                sent_count += 1
                log_notification('comment_added', task['id'], group['chat_id'], message, 'sent')
        
        return {'success': True, 'sent_to': sent_count}
    
    def handle_broadcast(self, data):
        """Handle broadcast message requests"""
        message_text = data.get('message')
        sender_name = data.get('sender_name', 'Admin')
        sender_email = data.get('sender_email', '')
        
        if not message_text:
            return {'error': 'No message provided'}
        
        message = (
            f"📢 <b>BROADCAST MESSAGE</b>\n"
            f"From: {sender_name}\n\n"
            f"{message_text}"
        )
        
        groups = get_active_groups()
        sent_count = 0
        failed_count = 0
        
        for group in groups:
            result = send_telegram_message(group['chat_id'], message)
            
            if result and result.get('ok'):
                sent_count += 1
                log_notification('broadcast', None, group['chat_id'], message, 'sent')
            else:
                failed_count += 1
                error_msg = result.get('description') if result else 'Unknown error'
                log_notification('broadcast', None, group['chat_id'], message, 'failed', error_msg)
        
        return {
            'success': True,
            'sent_to': sent_count,
            'failed': failed_count,
            'total_groups': len(groups)
        }
    
    def handle_deadline_reminder(self, data):
        """Send deadline reminders for tasks due soon"""
        from datetime import datetime, timedelta
        
        try:
            tomorrow = (datetime.utcnow() + timedelta(days=1)).isoformat()
            today = datetime.utcnow().isoformat()
            
            tasks_response = supabase.table('tasks')\
                .select('*')\
                .gte('deadline', today)\
                .lte('deadline', tomorrow)\
                .neq('status', 'completed')\
                .execute()
            
            if not tasks_response.data:
                return {'success': True, 'message': 'No upcoming deadlines'}
            
            groups = get_active_groups()
            total_sent = 0
            
            for task in tasks_response.data:
                deadline = datetime.fromisoformat(task['deadline'].replace('Z', '+00:00'))
                hours_left = int((deadline - datetime.utcnow()).total_seconds() / 3600)
                
                message = (
                    "⏰ <b>DEADLINE REMINDER</b>\n\n"
                    f"📋 <b>Task:</b> {task['title']}\n"
                    f"👤 <b>Assigned to:</b> {task['assigned_to']}\n"
                    f"⏱️ <b>Due in:</b> {hours_left} hours\n"
                    f"📅 <b>Deadline:</b> {task['deadline'][:10]}\n\n"
                    f"👉 <a href='{WEB_APP_URL}'>View Task</a>"
                )
                
                for group in groups:
                    result = send_telegram_message(group['chat_id'], message)
                    if result and result.get('ok'):
                        total_sent += 1
                        log_notification('deadline_reminder', task['id'], group['chat_id'], message, 'sent')
            
            return {
                'success': True,
                'tasks_checked': len(tasks_response.data),
                'notifications_sent': total_sent
            }
        except Exception as e:
            logger.error(f"Error in deadline reminder: {e}")
            return {'error': str(e)}

File: api/telegram-webhook.js
/**
 * Vercel Serverless Function for Telegram Webhook API
 * This handles Telegram notifications for the UmraGO Team Planner
 */

const { createClient } = require('@supabase/supabase-js');

// Configuration from environment variables
const TELEGRAM_BOT_TOKEN = process.env.TELEGRAM_BOT_TOKEN || '';
const SUPABASE_URL = process.env.SUPABASE_URL || '';
const SUPABASE_KEY = process.env.SUPABASE_ANON_KEY || '';
const WEB_APP_URL = process.env.WEB_APP_URL || 'https://team-planner-task-tracker.vercel.app/';
const WEBHOOK_SECRET = process.env.WEBHOOK_SECRET || '';

// Initialize Supabase client
const supabase = SUPABASE_URL && SUPABASE_KEY 
  ? createClient(SUPABASE_URL, SUPABASE_KEY)
  : null;

/**
 * Send a message via Telegram Bot API
 */
async function sendTelegramMessage(chatId, message) {
  try {
    const url = `https://api.telegram.org/bot${TELEGRAM_BOT_TOKEN}/sendMessage`;
    const response = await fetch(url, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        chat_id: chatId,
        text: message,
        parse_mode: 'HTML',
        disable_web_page_preview: true,
      }),
    });
    
    return await response.json();
  } catch (error) {
    console.error('Error sending Telegram message:', error);
    return null;
  }
}

/**
 * Get all active Telegram groups
 */
async function getActiveGroups() {
  try {
    if (!supabase) return [];
    
    const { data, error } = await supabase
      .from('telegram_groups')
      .select('*')
      .eq('is_active', true);
    
    if (error) throw error;
    return data || [];
  } catch (error) {
    console.error('Error getting active groups:', error);
    return [];
  }
}

/**
 * Log notification to database
 */
async function logNotification(notificationType, taskId, chatId, message, status = 'sent', error = null) {
  try {
    if (!supabase) return;
    
    const data = {
      notification_type: notificationType,
      task_id: taskId,
      chat_id: chatId,
      message_text: message,
      status: status,
    };
    
    if (error) {
      data.error_message = error;
    }
    
    await supabase.from('telegram_notifications').insert(data);
  } catch (error) {
    console.error('Error logging notification:', error);
  }
}

/**
 * Handle task created webhook
 */
async function handleTaskCreated(data) {
  const task = data.record;
  
  if (!task) {
    return { error: 'No task data provided' };
  }
  
  const message = (
    `🆕 <b>NEW TASK CREATED</b>\n\n` +
    `📋 <b>Title:</b> ${task.title}\n` +
    `👤 <b>Assigned to:</b> ${task.assigned_to}\n` +
    `📅 <b>Deadline:</b> ${task.deadline.substring(0, 10)}\n` +
    `⚡ <b>Priority:</b> ${task.priority.toUpperCase()}\n` +
    `✍️ <b>Created by:</b> ${task.created_by}\n\n` +
    `👉 <a href="${WEB_APP_URL}">View Task Details</a>`
  );
  
  const groups = await getActiveGroups();
  let sentCount = 0;
  let failedCount = 0;
  
  for (const group of groups) {
    const result = await sendTelegramMessage(group.chat_id, message);
    
    if (result && result.ok) {
      sentCount++;
      await logNotification('task_created', task.id, group.chat_id, message, 'sent');
    } else {
      failedCount++;
      const errorMsg = result?.description || 'Unknown error';
      await logNotification('task_created', task.id, group.chat_id, message, 'failed', errorMsg);
    }
  }
  
  return {
    success: true,
    sent_to: sentCount,
    failed: failedCount,
    total_groups: groups.length,
  };
}

/**
 * Handle task updated webhook
 */
async function handleTaskUpdated(data) {
  const task = data.record;
  const oldTask = data.old_record;
  
  if (!task) {
    return { error: 'No task data provided' };
  }
  
  const changes = [];
  if (oldTask) {
    if (task.status !== oldTask.status) {
      changes.push(`Status: ${oldTask.status} → ${task.status}`);
    }
    if (task.priority !== oldTask.priority) {
      changes.push(`Priority: ${oldTask.priority} → ${task.priority}`);
    }
  }
  
  if (changes.length === 0) {
    return { success: true, message: 'No significant changes' };
  }
  
  const message = (
    `🔄 <b>TASK UPDATED</b>\n\n` +
    `📋 <b>Task:</b> ${task.title}\n\n` +
    `<b>Changes:</b>\n${changes.map(c => `• ${c}`).join('\n')}\n\n` +
    `👉 <a href="${WEB_APP_URL}">View Task Details</a>`
  );
  
  const groups = await getActiveGroups();
  let sentCount = 0;
  
  for (const group of groups) {
    const result = await sendTelegramMessage(group.chat_id, message);
    if (result && result.ok) {
      sentCount++;
      await logNotification('task_updated', task.id, group.chat_id, message, 'sent');
    }
  }
  
  return { success: true, sent_to: sentCount };
}

/**
 * Handle task completed webhook
 */
async function handleTaskCompleted(data) {
  const task = data.record;
  
  if (!task || task.status !== 'completed') {
    return { error: 'Invalid task data' };
  }
  
  const message = (
    `✅ <b>TASK COMPLETED</b>\n\n` +
    `📋 <b>Task:</b> ${task.title}\n` +
    `👤 <b>Completed by:</b> ${task.assigned_to}\n` +
    `🎉 Great job!\n\n` +
    `👉 <a href="${WEB_APP_URL}">View Details</a>`
  );
  
  const groups = await getActiveGroups();
  let sentCount = 0;
  
  for (const group of groups) {
    const result = await sendTelegramMessage(group.chat_id, message);
    if (result && result.ok) {
      sentCount++;
      await logNotification('task_completed', task.id, group.chat_id, message, 'sent');
    }
  }
  
  return { success: true, sent_to: sentCount };
}

/**
 * Handle comment added webhook
 */
async function handleCommentAdded(data) {
  const comment = data.record;
  
  if (!comment) {
    return { error: 'No comment data provided' };
  }
  
  try {
    const { data: taskData, error } = await supabase
      .from('tasks')
      .select('*')
      .eq('id', comment.task_id)
      .single();
    
    if (error || !taskData) {
      return { error: 'Task not found' };
    }
    
    const message = (
      `💬 <b>NEW COMMENT</b>\n\n` +
      `📋 <b>Task:</b> ${taskData.title}\n` +
      `👤 <b>From:</b> ${comment.user_name}\n` +
      `💭 <b>Comment:</b> ${comment.comment_text}\n\n` +
      `👉 <a href="${WEB_APP_URL}">View Task</a>`
    );
    
    const groups = await getActiveGroups();
    let sentCount = 0;
    
    for (const group of groups) {
      const result = await sendTelegramMessage(group.chat_id, message);
      if (result && result.ok) {
        sentCount++;
        await logNotification('comment_added', taskData.id, group.chat_id, message, 'sent');
      }
    }
    
    return { success: true, sent_to: sentCount };
  } catch (error) {
    console.error('Error fetching task:', error);
    return { error: 'Failed to fetch task details' };
  }
}

/**
 * Main handler for Vercel serverless function
 */
module.exports = async (req, res) => {
  // Set CORS headers
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'POST, GET, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type, X-Webhook-Secret');
  
  // Handle OPTIONS (CORS preflight)
  if (req.method === 'OPTIONS') {
    return res.status(200).end();
  }
  
  // Handle GET (health check)
  if (req.method === 'GET') {
    if (req.url === '/api/telegram-webhook/health' || req.url === '/api/telegram-webhook') {
      return res.status(200).json({
        status: 'ok',
        service: 'UmraGO Telegram Webhook API',
      });
    }
    return res.status(404).json({ error: 'Not found' });
  }
  
  // Handle POST
  if (req.method === 'POST') {
    try {
      // Verify webhook secret
      const secret = req.headers['x-webhook-secret'];
      if (secret !== WEBHOOK_SECRET) {
        return res.status(401).json({ error: 'Unauthorized' });
      }
      
      const url = req.url;
      let result;
      
      // Route to appropriate handler
      if (url.includes('/task-created')) {
        result = await handleTaskCreated(req.body);
      } else if (url.includes('/task-updated')) {
        result = await handleTaskUpdated(req.body);
      } else if (url.includes('/task-completed')) {
        result = await handleTaskCompleted(req.body);
      } else if (url.includes('/comment-added')) {
        result = await handleCommentAdded(req.body);
      } else {
        return res.status(404).json({ error: 'Endpoint not found' });
      }
      
      return res.status(200).json(result);
    } catch (error) {
      console.error('Error handling request:', error);
      return res.status(500).json({ error: error.message });
    }
  }
  
  return res.status(405).json({ error: 'Method not allowed' });
};

